builtin: Implements ST_Segmentize builtin function

abhishek20123g opened the pull request:
Fixes: https://github.com/cockroachdb/cockroach/issues/48368

This PR adds ST_Segmentize({geography,float8}) builtin function
which allows modify given geography such that no segment longer
than the given max_segment_length.

Release note (sql change): This PR implements ST_Segmentize
builtin functions.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/49211)
<!-- Reviewable:end -->

abhishek20123g commented:
Small Overview:
1. Not sure its good idea to create the separate file as `geog_modification.go` 
   as all Other files are not suiting the definition of `ST_Segmentize`.
2. `SegmentizeGeography` and `segmentizeGeom` can be clubbed together but it was 
   easy to use recursion if they are separate.
3. `geog_modification_test.go` follow the same test case and structure as was in 
    `unary_operator_test.go`.
4. Since I am new to `S2` and `geom` library, tried to find a suitable function for interpolation or segmentization in `geom` library but unable to find suitable therefore it has to cast to `s2.point` as `s2` library consist function for interpolation.
5. Not sure the usage of `canUseIndex` in `infoBuilder` therefore set it to true as most of the relatable builtin have it as true.

otan commented on pkg/sql/sem/builtins/geo_builtins.go:
```diff
@@ -1067,6 +1067,36 @@ Note ST_Perimeter is only valid for Polygon - use ST_Length for LineString.`,
+			Info: infoBuilder{
+				info: "Returns a modified geometry having no segment longer than the given max_segment_length. " +
+					"Distance computation is performed in 2d only. Units are in meters.",
+				libraryUsage: usesGeographicLib | usesS2,
```
we don't actually use `GeographicLib` here, so just `libraryUsage: usesS2` is fine.

otan commented on pkg/sql/sem/builtins/geo_builtins.go:
```diff
@@ -1067,6 +1067,36 @@ Note ST_Perimeter is only valid for Polygon - use ST_Length for LineString.`,
+				return tree.NewDGeography(segGeometry), nil
+			},
+			Info: infoBuilder{
+				info: "Returns a modified geometry having no segment longer than the given max_segment_length. " +
```
`Returns a modified Geography having no segment longer than the given max_segment_length meters. ` +
`The calculations are done on a sphere.`

otan commented on pkg/geo/geogfn/geog_modification_test.go:
```diff
@@ -0,0 +1,153 @@
+		geogWkt: "LINESTRING(1.0 1.0, 2.0 2.0, 3.0 3.0)",
+		segmentize: segmentizedMaxLengthAndExpGeog{
+			maxSegmentLength: 100000.0,
+			segmentizedGeogWkt: "LINESTRING (1 1, 1.4998857365616758 1.5000570914791973, 2 2, 2.4998094835255658 2.500095075163195, 3 3)",
```
can we add tests for no segmentization needed for LineString / Polygon?

otan reviewed (commented):
Thank you for the contribution - this was harder than I expected and thanks for perservering!

I have a few comments that are mostly style related.

otan commented on pkg/geo/geogfn/geog_modification.go:
```diff
@@ -0,0 +1,145 @@
+// Copyright 2020 The Cockroach Authors.
```
nit also: you can name this segmentize.go / segmentize_test.go.

abhishek20123g commented:
**TF[YT]R**,
There seems to a lot of style-related review marks I need to be more careful :relaxed:.
Updated PR as you suggested.

abhishek20123g commented on pkg/geo/geogfn/geog_modification.go:
```diff
@@ -0,0 +1,145 @@
+
+	var segmentizedPoints []s2.Point
+	spheroid := geographiclib.WGS84Spheroid
+	distanceBetweenPoints := s2.ChordAngleBetweenPoints(point1, point2).Angle().Radians() * spheroid.SphereRadius
```
Done, 
That will be more optimized :blush:

abhishek20123g commented on pkg/geo/geogfn/geog_modification.go:
```diff
@@ -0,0 +1,145 @@
+	case *geom.MultiLineString:
+		segMultiLine := geom.NewMultiLineString(geom.XY).SetSRID(geometry.SRID())
+		for lineIdx := 0; lineIdx < geometry.NumLineStrings(); lineIdx++ {
+			l, _ := segmentizeGeom(geometry.LineString(lineIdx), segmentMaxLength)
```
Updated,
I just left error unhandled as `*geom.LinearRing` and `*geom.LineString` will never return.

otan commented on pkg/geo/geogfn/segmentize.go:
```diff
@@ -0,0 +1,161 @@
+	spheroid := geographiclib.WGS84Spheroid
+	// Convert segmentMaxLength to s1.ChordAngle as further calculation
+	// is done considering segmentMaxLength as s1.ChordAngle.
+	segmentMaxLength = segmentMaxLength / spheroid.SphereRadius
```
make this of type `s1.ChordAngle`:
`s1.ChordAngleFromAngle(s1.Angle(segmentMaxLength/spheroid.SphereRadius))`

and rename it to `segmentMaxChordAngle`

abhishek20123g replied:
This will be more efficient doing this way,
but I think `s1.ChordAngle` does not shows linear behaviour as
```
a := s1.ChordAngleFromAngle(s1.Angle(1))
b := s1.ChordAngleFromAngle(s1.Angle(0.5))

fmt.Println(a/b)
```
the expected ratio must be 2 but its actually 3.755165123780745
therefore, the ratio is not maintained after I convert them to `s1.ChordAngle`

otan replied:
Good find!

otan commented on pkg/geo/geogfn/segmentize.go:
```diff
@@ -0,0 +1,161 @@
+// the given maximum segment length.
+func segmentizeGeom(geometry geom.T, segmentMaxLength float64) (geom.T, error) {
+	if segmentMaxLength <= 0 {
+		return nil, errors.Newf("maximum segment length must be positive")
```
should this check be moved up to `Segmentize`?

abhishek20123g replied:
Sure,
This should be in `Segmentize` in the first place

otan commented on pkg/geo/geogfn/segmentize.go:
```diff
@@ -0,0 +1,161 @@
+		// (2^n)[numberOfSegmentToCreate] >= distanceBetweenPoints / segmentMaxLength > 2^(n-1)
+		// therefore n = ceil(log2(segmentMaxLength/distanceBetweenPoints)). Hence
+		// numberOfSegmentToCreate = 2^(ceil(log2(segmentMaxLength/distanceBetweenPoints))).
+		numberOfSegmentToCreate := int(math.Pow(2, math.Ceil(math.Log2(distanceBetweenPoints/maxSegmentLength))))
```
ok, this math is a little complicated. i think doing some inversing makes this cleaner:
```
numPoints := math.Ceil(maxSegmentLength / distanceBetweenPoints)
```
is there something wrong with doing it this way around?

abhishek20123g replied:
I think there may be a problem with directly writing it that way,
for example,
```
postgis_db=# select st_distance('POINT(0 0)'::geography, 'POINT(1 1)'::geography, false);
   st_distance   
-----------------
 157249.59776851
(1 row)

select st_astext(ST_Segmentize('LINESTRING(0 0, 1 1'::geography, 78624.798888))

postgis_db=# select st_astext(ST_Segmentize('LINESTRING(0 0, 1 1)'::geography, 78624.798888));
                        st_astext                        
---------------------------------------------------------
 LINESTRING(0 0,0.499961919922622 0.500019038226106,1 1)

postgis_db=# select st_astext(ST_Segmentize('LINESTRING(0 0, 1 1)'::geography, 78623.798888));
                                                            st_astext                                                            
---------------------------------------------------------------------------------------------------------------------------------
 LINESTRING(0 0,0.249976200223564 0.250011898653433,0.499961919922622 0.500019038226106,0.749966679297834 0.750016659002919,1 1)
```

for the case, one ratio will be 1.99999 
and the total number of segments was 2, and math.Ceil(distanceBetweenPoints/maxSegmentLength)
will as give 2.
and for case 2 ratio was 2.0002
and the total number of segments was 4, and math.Ceil(distanceBetweenPoints/maxSegmentLength)
will as give 3.

otan replied:
Looks like I misunderstood what this does. Thanks for setting me right.

otan commented on pkg/geo/geogfn/segmentize.go:
```diff
@@ -0,0 +1,161 @@
+	point2 := s2.PointFromLatLng(s2.LatLngFromDegrees(b.Y(), b.X()))
+
+	var segmentizedPoints []s2.Point
+	distanceBetweenPoints := s2.ChordAngleBetweenPoints(point1, point2).Angle().Radians()
```
because of the above transformation, we don't need to do `.Angle().Radians()`

otan commented on pkg/geo/geogfn/segmentize.go:
```diff
@@ -0,0 +1,161 @@
+// NOTE: List of points does not consist of end point.
+func segmentizeCoords(a geom.Coord, b geom.Coord, maxSegmentLength float64) []float64 {
+	// Converted geom.Coord into s2.Point so we can segmentize the coordinates.
+	point1 := s2.PointFromLatLng(s2.LatLngFromDegrees(a.Y(), a.X()))
```
nit: pointA, pointB.

otan commented on pkg/geo/geogfn/segmentize.go:
```diff
@@ -0,0 +1,161 @@
+			segmentizedPoints = append(segmentizedPoints, newPoint)
+		}
+	}
+	allSegmentizedCoordinates := a.Clone()
```
seems like we're iterating over the array twice, and allocate two arrays twice. can we restructure this to look like:

```
segmentizedCoordinates := []float64{}
segmentizedCoordinates = append(segmentizedCoordinates, a.Clone())
if maxSegmentLength <= distanceBetweenPoints {
   // ....
   for pointInserted := 1; pointInserted < numberOfSegmentToCreate; pointInserted++ {
      nextPoint := s2.Interpolate(....)
      segmentizedCoordinates = append(segmentizedCoordinates, nextPoint.Lng.Degrees(), latlng.Lat.Degrees())
  }
}
return segmentizedCoordinates
```

otan commented on pkg/sql/sem/builtins/geo_builtins.go:
```diff
@@ -1129,6 +1129,36 @@ Note ST_Perimeter is only valid for Polygon - use ST_Length for LineString.`,
+				return tree.NewDGeography(segGeometry), nil
+			},
+			Info: infoBuilder{
+				info: `turns a modified Geography having no segment longer than the given max_segment_length meters. 
```
nit: replace "turns" with "Returns". Extra new line after "meters." so that they appear in the autogenerated markdown as two different <p> tags.

otan commented on pkg/sql/sem/builtins/geo_builtins.go:
```diff
@@ -1129,6 +1129,36 @@ Note ST_Perimeter is only valid for Polygon - use ST_Length for LineString.`,
+		tree.Overload{
+			Types: tree.ArgTypes{
+				{"geography", types.Geography},
+				{"max_segment_length", types.Float},
```
late realisation: maybe renaming this to `max_segment_length_meters` helps. now we can omit the "meters" word in the description!

(spelling metres as meters is painful for an aussie but here we are)

abhishek20123g replied:
:laughing:

otan commented on pkg/geo/geo.go:
```diff
@@ -188,6 +188,15 @@ func NewGeography(spatialObject geopb.SpatialObject) *Geography {
 	return &Geography{SpatialObject: spatialObject}
 }
 
+// NewGeographyFromGeom Geography Object from geom.T.
```
nit: "NewGeographyFromGeom creates a new Geography from a geom.T object."

abhishek20123g replied:
Done.

abhishek20123g commented:
Late to realise that for `POINT` and `MULTIPOINT`, `St_Segmentize` works with 0 and negative segment length.
```
postgis_db=# SELECT ST_AsText(ST_Segmentize('MULTIPOINT(0 0, 1 1)'::geography, -1));
      st_astext      
---------------------
 MULTIPOINT(0 0,1 1)
```
```
postgis_db=# SELECT ST_AsText(ST_Segmentize('LINESTRING(0 0, 1 1)'::geography, -1));
ERROR:  ptarray_segmentize_sphere: maximum segment length must be positive
```
updated PR according to this case and added their respective testcase in 
both `segmentize_test.go` and `geospatial` file

otan commented:
heads up this looks good to me - there's a few tests i want to add but they won't work yet because i have to land the patches for making them work (if you're curious, its strange types like `LINESTRING EMPTY`). i'll update your PR with this tomorrow and submit it. thanks for your contribution.

otan committed:
builtin: implements ST_Segmentize builtin function

Fixes: https://github.com/cockroachdb/cockroach/issues/48368

This PR adds ST_Segmentize({geography,float8}) builtin function
which allows modify given geography such that no segment longer
than the given max_segment_length.

Co-authored-by: abhishek20123g <abhishek20123g@gmail.com>

Release note (sql change): This PR implements the ST_Segmentize
builtin function for Geography.

otan commented:
ok! this is ready to go. feeling like an extension? you can do this for geometry as well :P!
https://github.com/cockroachdb/cockroach/issues/49029

bors r+

otan commented:
oops, lost your name in the commit, one sec

bors r-

otan commented:
bors r+
